manifests/0000_90_kube-apiserver-operator_04_servicemonitor-apiserver.yaml

```
@@ -135,3 +135,16 @@ spec:
     - record: cluster:apiserver_current_inflight_requests:sum:max_over_time:2m
       expr: |
         max_over_time(sum(apiserver_current_inflight_requests{apiserver=~"openshift-apiserver|kube-apiserver"}) by (apiserver,requestKind)[2m:])
+  - name: pre-release-lifecycle
+    rules:
+      - alert: DeprecratedAPIUsed
+        annotations:
+          message: >-
+            Deprecated API that will be removed in the next version is being used. Removing the workload that is using
+            the {{"{{$labels.group}}"}}.{{"{{$labels.version}}"}}/{{"{{$labels.resource}}"}} API might be necessary for
+            a successful upgrade to the next cluster version.
+        expr: |
+          group(apiserver_requested_deprecated_apis{removed_release="1.21"}) by (group,version,resource,subresource) * on(group,version,resource,subresource) group_right() rate(apiserver_request_total[10m]) > 0
```

how do we make sure this version is updated each release?

---

If you want to you can either edit for the next release so in 4.8 you would replace or add an or for 1.22, or use a regex matcher here that would match against multiple versions, see -> https://prometheus.io/docs/prometheus/latest/querying/basics/#time-series-selectors

---

my question was more about how we remember this in practice. We could add CI test that compares the value with the running server version.

---

Yes that is a good idea, or something along the lines of creating one of the deprecated APIs and evaluating if the alert is firing.

---

@sttts Added an e2e test.

---

Suggestion to get rid of the `on(...) group_right()` part

```suggestion
          group(apiserver_requested_deprecated_apis{removed_release="1.21"}) by (group,version,resource,subresource) and (sum by(group,version,resource,subresource) (rate(apiserver_request_total[10m]))) > 0
```

You can even remove the `subresource` label from the by() clauses since it's not used in the annotations.

---

@simonpasquier Thank you. Updated.